grpc/noncebalancer/noncebalancer.go

```diff
@@ -0,0 +1,105 @@
+package noncebalancer
+
+import (
+	"errors"
+
+	"github.com/letsencrypt/boulder/nonce"
+	"google.golang.org/grpc/balancer"
+	"google.golang.org/grpc/balancer/base"
+)
+
+// Compile-time assertion that *Picker implements the base.PickerBuilder
+// interface.
+var _ base.PickerBuilder = (*Balancer)(nil)
+
+var errNoPrefix = errors.New("nonce.PrefixKey value required in RPC context")
+var errNoPrefixSalt = errors.New("nonce.PrefixSaltKey value required in RPC context")
+var errPrefixType = errors.New("nonce.PrefixKey value in RPC context must be a string")
+var errPrefixSaltType = errors.New("nonce.PrefixSaltKey value in RPC context must be a string")
+
+// Balancer is a base.Balancer used to construct a new Picker. It implements the
+// base.PickerBuilder interface but should only be used as a base.Balancer for
+// nonce server clients. It can be invoked by passing
+// `{"loadBalancingConfig":[{"nonce":{}}]}` as the default service config to
+// grpc.Dial().
```

Comment:
I assume is referring to [`grpc.WithDefaultServiceConfig`](https://pkg.go.dev/google.golang.org/grpc#WithDefaultServiceConfig)? That seems like an option for handling fallbacks, when the name resolver doesn't provide a service config.

In our case, what we want is not a fallback, but something mandatory, right? A service that is configured to connect to backends using the nonce service approach should always use the nonce Balancer.

Is this a way to avoid having to define a separate name resolver that returns a special service config for the nonce load balancer? If so, that's fine, but we should note why we're doing it.

Reply:
There are four ways to specify a Load Balancer implementation:

1. Don't. In this case the `grpc.ClientConn` will default to [`pick_first`](https://github.com/grpc/grpc-go/blob/v1.52.3/pickfirst.go).
1. [grpc.WithDefaultServiceConfig](https://pkg.go.dev/google.golang.org/grpc#WithDefaultServiceConfig) if we know that our `grpc.Resolver` won't override it with an undesirable implementation.
1. [grpc.WithServiceConfig](https://pkg.go.dev/google.golang.org/grpc#WithServiceConfig), which is deprecated in favor of either `grpc.WithDefaultServiceConfig` or passing the balancer name via your own `grpc.Resolver`.
1. A custom `grpc.Resolver` implementation which overrides "`pick_first`" and/or the Balancer specified by `grpc.WithDefaultServiceConfig`.

We never want to default to `"pick_first"`, so we pass `"round_robin"` in our RPC client constructor. We're using our own SRV resolver implementation, but it's generalized for several RPC clients. In the end I settled on `grpc.WithDefaultServiceConfig`.

Reply:
Yep, that makes sense. To me it seems like (4) is the best, blessed way to do things, and (2) is surprising. In particular, what if we later change our SRV resolver implementation so it specifies a Load Balancer (e.g. `least_loaded`)? We would break the nonce balancer even though there's isn't really a clear connection between them.

What about giving our SRV dnsResolver a field that indicates which balancer we want to use? We would embed that into two different structs: one which has a Scheme of "srv" and uses `round_robin`, and one which has a Scheme of "srv_nonce" and uses `nonce`?

Reply:
Okay, I hadn't thought about registering it twice as two different load balancers. That's a wonderful idea!

Reply:
s/load balancers/resolvers/